src/env-inl.h

```diff
@@ -355,18 +343,19 @@ Environment::BindingScope<T>::~BindingScope() {
 }
 
 template <typename T>
-v8::MaybeLocal<v8::Object> Environment::MakeBindingCallbackData() {
+v8::MaybeLocal<v8::External> Environment::MakeBindingCallbackData(
+    void* allocation) {
   v8::Local<v8::Function> ctor;
   v8::Local<v8::Object> obj;
   if (!as_callback_data_template()->GetFunction(context()).ToLocal(&ctor) ||
       !ctor->NewInstance(context()).ToLocal(&obj)) {
-    return v8::MaybeLocal<v8::Object>();
+    return v8::MaybeLocal<v8::External>();
   }
-  T* data = new T(this, obj);
+  if (allocation == nullptr) allocation = operator new(sizeof(T));
```

Comment:
Why do we do this here instead of using `DeserializeInternalFieldsCallback`?

Reply:
@joyeecheung How would that look like?

Reply:
IIUC `BaseObject::SerializeInternalFields` here are just noops (other than they perform some checks...in addition to the checks that V8 already performs), even though the V8 API is designed to save the association between the JS value (externals here) and the embedder fields. I think we are bypassing the entire mechanism here by assuming that the index returned by V8 starts from zero and thus matching our internal indices of the referenced payload here? Although if we play along with the API, the payload should be retrieved and reinterpreted from the `DeserializeInternalFieldsCallback` as serialized in in the `SerializeInternalFieldsCallback` passed into `AddContext`, and what you need to save additionally to the snapshot data are just indices, instead of payloads being read/written using the low-level readers/writers here (V8 will just find the payload in the blob for you when given indices) - so ideally the readers and writers should only have methods for writing and reading `int`s.

Reply:
> IIUC `BaseObject::SerializeInternalFields` here are just noops (other than they perform some checks...in addition to the checks that V8 already performs),

Yes, that’s correct.

>  even though the V8 API is designed to save the association between the JS value (externals here) and the embedder fields.

It’s not really clear to me what this means. I thought the V8 internal fields serialization/deserialization hooks were designed to save the association between the JS object and the native object(s) that the internal fields refer to?

> I think we are bypassing the entire mechanism here by assuming that the index returned by V8 starts from zero and thus matching our internal indices of the referenced payload here? 

No, nothing in this codes makes such assumptions. 

> Although if we play along with the API, the payload should be retrieved and reinterpreted from the `DeserializeInternalFieldsCallback` as serialized in in the `SerializeInternalFieldsCallback` passed into `AddContext`,

Right, we could add a callback for that if we ever run into a case in which the internal fields cannot be set properly otherwise.

Currently, there is no need to do so, because all objects with internal fields that are being serialized (and all objects that I could imagine will be added to the snapshot in the future) are `BaseObject`s. Since `BaseObject`s are V8 persistent handles tied together with C++ objects, and we have to serialize the persistent handles anyway, we already have one serialization step that performs the necessary storage of data during the “serialize all global handles” phase. When deserializing, we know how to re-create the `BaseObject` based on the deserialized data, and the internal fields will be set as they usually would during regular program execution.

This is why the `BaseObject` internal fields callback serializer only is a direct no-op for the first internal field – the first one is one where we *know* that the `BaseObject` constructor resets it to a proper value, and where we know it is semantically valid to do so. For other potential fields, that’s not a given.

> and what you need to save additionally to the snapshot data are just indices, instead of payloads being read/written using the low-level readers/writers here (V8 will just find the payload in the blob for you when given indices) - so ideally we should only use readers/writers for `int`s when serializing/deserializing these values (and reserve the other types for special data serialized along with the Environment itself that are not associated with any JS values)

Tbh, I don’t really understand what you are referring to here. It would be good if you could expand on that.

Reply:
> > and what you need to save additionally to the snapshot data are just indices, instead of payloads being read/written using the low-level readers/writers here (V8 will just find the payload in the blob for you when given indices) - so ideally we should only use readers/writers for `int`s when serializing/deserializing these values (and reserve the other types for special data serialized along with the Environment itself that are not associated with any JS values)
> 
> Tbh, I don’t really understand what you are referring to here. It would be good if you could expand on that.

It just occurred to me that you’re probably referring to the kind of index returned by `SnapshotCreator::AddData()`/passed to `GetDataFromSnapshotOnce()`. :slightly_smiling_face: For those, the classes that are being serialized, like `Environment`, don’t currently have to deal with passing the index around, and instead leave that up to the snapshot writer/reader API in this PR (look around for `V8SnapshotIndex`, I’ve added a separate typedef for that in the last fixups commit to be more explicit about what kind of indices those are).

Reply:
> When deserializing, we know how to re-create the BaseObject based on the deserialized data, and the internal fields will be set as they usually would during regular program execution.

I was talking about instead of deserializing the native states in BaseObject as if they are initialized during regular program execution (with out-of-band data coming from our own snapshot blob), use the `DeserializeInternalFieldsCallback`  and retrieve the native states from the `StartupData` payload - so effective when we are storing, say, a subclass of `BaseObject` with some states in the types of ints and bools, we don't use the `snapshot_data->Write*` methods during serialization to save them into the blob managed by us. Instead we just put them into a plain object and pass it to the `SerializeInternalFieldsCallback` invoked on the `BaseObject` (these will be invoked because the `Object` associated with `BaseObject` has embedder fields), so these states will be stored inside the blob managed by V8. At serialization these blobs will just pop up in the `DeserializeInternalFieldsCallback` and we'd use the information we stored to allocate the native memory pointed to by the `BaseObject` and initialize those native states.

Reply:
@joyeecheung Okay, what’s the advantage of that compared to the current approach? That seems like it would end up using even lower-level read/write operations, something that you explicitly seem to want to avoid.

Reply:
> That seems like it would end up using even lower-level read/write operations

In which situation? If we make use of the callbacks, most read/writes should be done by V8 and we'd be just passing things to it by value and casting blobs it pass to us?

Reply:
@joyeecheung Are you suggesting that we take the memory underlying a `BaseObject` and pass that along as the `StartupData` returned from the internal fields serialization callback? That’s bound to go wrong as soon as there’s any type of pointer in there, so I assume that’s not what you’re referring to?

Reply:
@addaleax I am talking about collecting values that can be recreated and pass them into the `StartupData`. When the states involve values, we collect those values, when there are pointers, we'd want to collect information on how to resurrect that pointer instead. The thing being passed to `StartupData` would not be exactly the same thing as pointed to by the `BaseObject` e.g. even if the memory being pointed to by the BaseObject is just a plain object with owned values, we may want to save an additional enum denoting what this BaseObject is to help debugging. The layout of this `StartupData` blob would be somewhat similar to the chunk we currently Write* into our own snapshot data in this PR, just that they will be collected by value and pass as a `StartupData` to V8, and V8 will be in charge of writing them into its own blob. For example, for this chunk in our own blob here

```
5175   StartEntry: [BaseObjects]
5197     StartEntry: [NoBindingData]
5221       Object index: 32
5231       EndEntry
5232     EndEntry
```

What we should essentially do is creating a blob like this:

```
class StartupDataForNoBindingData {
  Tag tag;  // a enum with NoBindingData
  V8SnapshotIndex index;  // collected as 32 in the serialization callback
}
```

And it will be stored by V8, instead of by us.

Reply:
@joyeecheung Why is it important that it’s stored separately, rather than “by us”? I think it would be fairly straightforward to create another `SnapshotCreateData` instance, write the data into it, and then return that as the `StartupData`, but is there any advantage? It sounds like letting V8 store the values in its own snapshot would make things harder to inspect for debugging.

> And pass it into the Deserialize callback so it will be stored by V8, instead of by us. In the serialization callback, we'll know the type of this blob by inspecting the type first, and allocate sizeof(NoBindingData) for it, then deserialize the object with index 32 and point the persistent handle to it.

I’m not sure that I understand why we would allocate data instead of directly creating the `BaseObject`, like it is currently performed? Or is your concern specifically about `NoBindingData`?

Reply:
> I’m not sure that I understand why we would allocate data instead of directly creating the BaseObject, like it is currently performed?

Hm, right, the order is wrong, we should deserialize the object with index 32 first, and then just make a call to `new NoBindingData(env, object)`.

> Why is it important that it’s stored separately, rather than “by us”? 

They should then be stored alongside with the JS values by V8, so not actually "separately" (on the other hand, we are currently storing the native states in our own blob and the JS states in V8's blob so I think this would count as "separately" instead?)

> It sounds like letting V8 store the values in its own snapshot would make things harder to inspect for debugging.

The thing is then we don't need to maintain the reading and writing of these blobs at all. We only need to care about how to collect these values. As a result our own blob would be significantly smaller.

Reply:
> > It sounds like letting V8 store the values in its own snapshot would make things harder to inspect for debugging.
> 
> The thing is then we don't need to maintain the reading and writing of these smaller blobs at all. We only need to care about how to collect these values.

I don’t see how this would affect the amount of serialization code that we use, tbh.

> As a result our own blob would be significantly smaller (reduced to mostly indices for recreating references from C++ to JS, and states that are not embedded as part of any JS value)

Right, but it makes the V8 snapshot larger by the same amount.

Given that this would make debugging harder, I’m inclined to keep the code simple instead, unless there’s a concrete advantage to using the internal fields serialization callback.

Reply:
> I don’t see how this would affect the amount of serialization code that we use, tbh.

Because we then process things by value, instead of by procedures? I find it easier to understand "what should I assign to in blob.index, which will come back the same way when I reinterpret the blob at deserialization", instead of wondering "what should I write into a blob as an index, which I will need to remember to read again as an index later in the same order I write them".

> Given that this would make debugging harder

How would this make debugging harder, though? These blobs will have pre-defined structures that can be inspected by name (either in the debugger, or in the callback). If it's talking about the ability to inspect all the native states in the entire snapshot, I am not sure how much saving them all in our own blob would help, since the corresponding JS states are always buried in V8's blob and most things we see are indices. For native states detached from JS values, they are still inspectable since we'll still save them in our own blob since there will be no callbacks invoked on them - and they are probably more likely to be the things we care about when we want to dump the snapshot. Mismatches of these smaller blobs are guarded by V8's snapshot creator implementation and we should not worry about them, if we don't trust V8's ability to pop these blobs back to us, we should be improving V8's API instead of creating our own.

> I’m inclined to keep the code simple instead

Why is this simpler than letting V8 read/write the values and invoke the callbacks for us? For example, we are maintaining our own maps of BaseObjectDeserializers as a result?

Reply:
> Right, but it makes the V8 snapshot larger by the same amount.

BTW, I don't think it does - with the current approach, we need one tag per field, plus start and end entries, as well as class name strings that aren't shared across these blobs. With predefined structures, we just need one tag per StartupData (which can include many fields, and there is no need to further tag these fields), and the structural information won't be proportional to the number of objects in the snapshots (they'll be in DWARF).

Reply:
> > I don’t see how this would affect the amount of serialization code that we use, tbh.
> 
> Because we then process things by value, instead of by procedures? I find it easier to understand "what should I assign to in blob.index, which will come back the same way when I reinterpret the blob at deserialization", instead of wondering "what should I write into a blob as an index, which I will need to remember to read again as an index later in the same order I write them".

I’m really not sure what you mean here. You have to write the `v8::Global`’s index either way, and you don’t need to maintain a specific order of deserialization manually, either.

> > Given that this would make debugging harder
> 
> How would this make debugging harder, though? These blobs will have pre-defined structures that can be inspected by name (either in the debugger, or in the callback). If it's talking about the ability to inspect all the native states in the entire snapshot, I am not sure how much saving them all in our own blob would help, since the corresponding JS states are always buried in V8's blob and most things we see are indices.

It would remove the `BaseObject` serializations from the serialization printout, for example – both with what this PR suggests and your plain C++ object suggestion. Yes, the correponding JS values are stored on V8’s side and not easily accessible, but even knowing what types of `BaseObject`s we have stored, how many, and what they look like on the native side seems very useful to me.

> > I’m inclined to keep the code simple instead
> 
> Why is this simpler than letting V8 read/write the values and invoke the callbacks for us? 

I could ask the converse question – why does it seem simpler for you to create a serialization structure, then write that/read that from the V8 snapshot?

> For example, we are maintaining our own maps of BaseObjectDeserializers as a result?

What else would you do? Maintain a list of all possible `BaseObject` types? That’s a code smell, imo. We should not be maintaining a global list of those.

Reply:
> I’m really not sure what you mean here. You have to write the v8::Global’s index either way, and you don’t need to maintain a specific order of deserialization manually, either.

I am talking about the native states of a C++ object wrapped inside a JS object (i.e. the internal fields). For example if the C++ object (likely a BaseObject) contains a bool, a uint32_t, and a string, with the current approach we need to invoke `StartWriteEntry`, `WriteBool`, `WriteUint32`, `WriteString` and `EndWriteEntry`, and during deserialization, we need to do the Reads in the exact same order to restore states for this C++ object. Using V8's callback, we will predefined a structure, assign fields to an instance of that accordingly (the order would not matter) and return it as a StartupData during serialization. At deserialization we just need to reinterpret the StartupData passed into the callback, and restore the states into the C++ object with information in that StartupData mostly by assigning fields, again the order does not matter - no error should occur if we assign the string first (probably by invoking the string constructor with a length and bytes), then the uint32 later.

> I could ask the converse question – why does it seem simpler for you to create a serialization structure, then write that/read that from the V8 snapshot?

As explained above, the order would not matter - the fields are identified by name inside that blob, and they will be put into place during deserialization, we just need to cast them, instead of reading them one-by-one from the blob in a specific order.

> Maintain a list of all possible BaseObject types? That’s a code smell, imo. We should not be maintaining a global list of those.

Why would that be a code smell? We already do that for AsyncWraps, it doesn't seem so bad to extend that to other subclasses of BaseObjects. We could return this in `MemoryInfoName()`, that gets converted to strings by the `MemoryTracker`, which helps us make sure that all names like this have a proper declaration (ideally, if we put a documentation about all existing BaseObject types above that declaration of types, I think it would be tremendously helpful for people to understand the structure of the code base), and unify the class names in both the heap snapshot and the startup snapshot. Also theoretically, if we put that type directly into the BaseObject, tools like llnode can also reuse that field for interpreting internal fields inside objects associated with `BaseObject`s - llnode was made possible because V8 has such global list of heap object types.

Reply:
Also to be clear, I don't think we have to use the V8 callbacks as-is, but I think we should go over why they can't be used for our use cases, and ideally, if they don't work for us, we can improve the API to make them more feasible, instead of bypassing the entire machinery and using our home-grown ones in an uncoordinated way.

Reply:
> > Maintain a list of all possible BaseObject types? That’s a code smell, imo. We should not be maintaining a global list of those.
> 
> Why would that be a code smell? We already do that for AsyncWraps, it doesn't seem so bad to extend that to other subclasses of BaseObjects.

Yes, also a code smell for `AsyncWrap`s, imo – it’s just a bit hard to get rid of that given that we store the names for them in `IsolateData`.

> We could return this in `MemoryInfoName()`

I don’t think it’s a good idea to tie identifiers and debugging names together, tbh. There’s always a possibility that we want to provide extra information in `MemoryInfoName()` that we wouldn’t need for identifying the type.